### PR TITLE
Typescript port

### PR DESCRIPTION
**Pull request by Bettelstab**
Ports code to Typescript. @ts-ignores are only in parts which should be removed eventually (related to https://github.com/nimiq-network/core/pull/484)

### CODE REVIEW

**Comment by danimoh** [comment]
Does Nimiq.d.ts need to be copied to this project or could it just be an npm dependency?

File: package.json [hunk 1]
```diff
@@ -1,22 +1,26 @@
...
+    "dist",
+    "types/index.d.ts"
   ],
   "dependencies": {
-    "@nimiq/utils": "^0.0.2"
+    "@nimiq/utils": "^0.2.0"
   },
   "devDependencies": {
-    "rollup": "^0.65.0"
+    "@nimiq/core-types": "^1.4.2",
```
Comment (sisou) [inline package.json:22]: Core-types is deprecated. You need to use the namespace from core-web.

File: package.json [hunk 2]
```diff
@@ -1,22 +1,26 @@
...
+    "rollup": "^0.65.0",
+    "typescript": "^3.3.3333"
```
Comment (sisou) [inline package.json:24]: What kind of version is this?

**Review by sisou — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (Bettelstab) [reply package.json:22]: But https://github.com/nimiq-network/core/pull/479 :(

File: package.json (hunk 2 above)
Comment (Bettelstab) [reply package.json:24]: https://github.com/Microsoft/TypeScript/issues/30032

> I would really like to know how many developer hours have been wasted worldwide investigating about this version number... with great power comes great responsibility :)
--

**Comment by Bettelstab** [comment]
> Does Nimiq.d.ts need to be copied to this project or could it just be an npm dependency?

Edit: Ooops. You meant *that* file. Don't know how it sneaked in. We don't need it.

File: src/NanoApi.ts [hunk 3]
```diff
@@ -0,0 +1,695 @@
+import { Utf8Tools } from '@nimiq/utils';
+import { TransactionReceipt } from '@nimiq/core-web';
```
Comment (sisou) [inline src/NanoApi.ts]: I think `@nimiq/core-web` is not in the dependencies, so does this even build?

File: package.json [hunk 4]
```diff
@@ -1,22 +1,26 @@
...
+    "dist",
+    "types/index.d.ts"
   ],
   "dependencies": {
-    "@nimiq/utils": "^0.0.2"
+    "@nimiq/utils": "^0.2.0"
   },
   "devDependencies": {
-    "rollup": "^0.65.0"
+    "@nimiq/core-types": "^1.4.3",
```
Comment (sisou) [inline package.json]: `1.4.3` does not exist (yet).

**Review by sisou — CHANGES_REQUESTED** [review]
Fails to build with
```
src/NanoApi.ts:343:43 - error TS2345: Argument of type 'Hash | null' is not assignable to parameter of type 'Serializable'.
  Type 'null' is not assignable to type 'Serializable'.
```

**Review by sisou — APPROVED** [review]
When you can assure that the NanoApi still works with the Safe and the Accounts Manager, then this is good to go.

File: package.json (hunk 2 above)
Comment (danimoh) [reply package.json:24]: Trolling at its best :rofl:

**Comment by Bettelstab** [comment]
Hmm. It breaks compatibility with accounts, as it removes the pico replacements. So better not merge it yet.

**Comment by Bettelstab** [comment]
Glad to see it was taken care of doing the port eventually :)

**Pull request closed without merging**